L1Trigger/L1TTrackMatch/plugins/L1TkMuonProducer.cc

```diff
@@ -0,0 +1,712 @@
+
+  explicit L1TkMuonProducer(const edm::ParameterSet&);
+  ~L1TkMuonProducer() override;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  PropState propagateToGMT(const L1TTTrackType& l1tk) const;
+  double sigmaEtaTP(const RegionalMuonCand& mu) const;
+  double sigmaPhiTP(const RegionalMuonCand& mu) const;
+
+  // the TP algorithm
+  void runOnMTFCollection_v1(const edm::Handle<RegionalMuonCandBxCollection>&,
+                             const edm::Handle<L1TTTrackCollectionType>&,
+                             TkMuonCollection& tkMuons,
+                             const int detector) const;
+
+  // algo for endcap regions using dynamic windows for making the match
+  void runOnMTFCollection_v2(const edm::Handle<EMTFTrackCollection>&,
+                             const edm::Handle<L1TTTrackCollectionType>&,
+                             TkMuonCollection& tkMuons) const;
+
+  // given the matching indexes, build the output collection of track muons
+  void build_tkMuons_from_idxs(TkMuonCollection& tkMuons,
+                               const std::vector<int>& matches,
+                               const edm::Handle<L1TTTrackCollectionType>& l1tksH,
+                               const edm::Handle<RegionalMuonCandBxCollection>& muonH,
+                               int detector) const;
+
+  // as above, but for the TkMu that were built from a EMTFCollection - do not produce a valid muon ref
+  void build_tkMuons_from_idxs(TkMuonCollection& tkMuons,
+                               const std::vector<int>& matches,
+                               const edm::Handle<L1TTTrackCollectionType>& l1tksH,
+                               int detector) const;
+
+  // dump and convert tracks to the format needed for the MAnTra correlator
+  std::vector<L1TkMuMantraDF::track_df> product_to_trkvec(const L1TTTrackCollectionType& l1tks);  // tracks
+  std::vector<L1TkMuMantraDF::muon_df> product_to_muvec(
+      const RegionalMuonCandBxCollection& l1mtfs);  // regional muon finder
+  std::vector<L1TkMuMantraDF::muon_df> product_to_muvec(
+      const EMTFTrackCollection& l1mus);  // endcap muon finder - eventually to be moved to regional candidate
+
+  float ETAMIN_;
+  float ETAMAX_;
+  float ZMAX_;  // |z_track| < ZMAX in cm
+  float CHI2MAX_;
+  float PTMINTRA_;
+  float DRmax_;
+  int nStubsmin_;  // minimum number of stubs
+  bool correctGMTPropForTkZ_;
+  bool use5ParameterFit_;
+  bool useTPMatchWindows_;
+
+  AlgoType bmtfMatchAlgoVersion_;
+  AlgoType omtfMatchAlgoVersion_;
+  AlgoType emtfMatchAlgoVersion_;
+
+  std::unique_ptr<L1TkMuCorrDynamicWindows> dwcorr_;
+
+  std::unique_ptr<L1TkMuMantra> mantracorr_barr_;
+  std::unique_ptr<L1TkMuMantra> mantracorr_ovrl_;
+  std::unique_ptr<L1TkMuMantra> mantracorr_endc_;
+  int mantra_n_trk_par_;
+
+  const edm::EDGetTokenT<RegionalMuonCandBxCollection> bmtfToken;
```

Comment:
member variables should end with underscore

Reply:
done